### PR TITLE
Run Validation locally

### PR DESCRIPTION
**Pull request by bnmnetp**
replace calling the w3 service with running the validation locally.

- a bit ugly since everything that uses the w3 validator (vnu.jar) requires java!
- Errors are captured on stdout

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -37,15 +38,16 @@
...
 import six
 from six.moves.urllib.error import HTTPError, URLError
 from six.moves.urllib.request import urlopen
+from html5validator.validator import Validator
 
 # Local imports
 # -------------
 from .utils import COVER_DIRS, DictToObject
 
+W3_VALIDATE = True
```
Comment (bjones1) [inline tests/conftest.py]: I'd recommend adding a brief comment to explain what this does. Another handy option would be to add a command-line flag to disable it.

File: tests/conftest.py [hunk 2]
```diff
@@ -347,23 +349,16 @@ def validate(self,
...
-                        if vld.warnings:
-                            print('Warnings for {}: {}'.format(url, len(vld.warnings)))
-                            pprint(vld.warnings)
-                        done = True
-                    except HTTPError as e:
-                        count -= 1
+            if expected_errors is not None and W3_VALIDATE:
+                # Redo this section using html5validate command line
+                vld = Validator(errors_only=True)
+                tmpname = 'tmphtml.html'
```
Comment (bjones1) [inline tests/conftest.py]: I would recommend using [tmp_path](http://doc.pytest.org/en/latest/tmpdir.html) to create a directory in which to store `tmphtml.html`. I'd rather not have temp files in the local path, even if the code removes them. (Note that pytest will eventually delete these directories, per https://github.com/pytest-dev/pytest/issues/1120).

File: tests/conftest.py [hunk 3]
```diff
@@ -347,23 +349,16 @@ def validate(self,
...
+                with open(tmpname, 'w') as f:
+                    f.write(self.text)
+                errors = vld.validate([tmpname])
+                os.remove(tmpname)
```
Comment (bjones1) [inline tests/conftest.py]: Remove. Pytest will take care of this.

File: tests/conftest.py (hunk 2 above)
Comment (bnmnetp) [reply tests/conftest.py]: I didn't see it as any different than what the other validator did with leaving files in the web2py root folder.  If they are temporary and pytest cleans them up then why add the extra code to put them somewhere else??

File: tests/conftest.py (hunk 1 above)
Comment (bnmnetp) [reply tests/conftest.py]: Yep.

File: tests/conftest.py (hunk 2 above)
Comment (bnmnetp) [reply tests/conftest.py]: Although now that I look at the docs, it seems that tmp_path / 'tmpfile.html'   Is probably a net zero change.

Comment (bnmnetp) [reply tests/conftest.py]: but no idea how you incorporate a fixture into code inside a class in conftest.py

Comment (bjones1) [reply tests/conftest.py]: I agree that my earlier code putting files in the web2py root folder is sloppy and should also be changed.

To incorporate this, my sketch of the code is:

```
# Create a client for accessing the Runestone server.
class _TestClient(WebClient):
    def __init__(self, web2py_server, web2py_server_address, runestone_name, tmp_path <-- added this):
        self.web2py_server = web2py_server
        self.web2py_server_address = web2py_server_address
        self.tmp_path = tmp_path # <-- added this
```
Then modify the fixture just after it to take `tmp_path` and pass it to the constructor.

**Pull request merged**